File: app/src/main/java/com/tomclaw/drawa/tools/Eraser.java
package com.tomclaw.drawa.tools;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;

import com.tomclaw.drawa.DrawHost;

/**
 * Created by solkin on 17.03.17.
 */
public class Eraser extends Radiusable {

    private int startX, startY;
    private int prevX, prevY;
    private Path path;
    private int externalColor;

    public Eraser() {
    }

    @Override
    public void onInitialize() {
        this.path = new Path();
    }

    @Override
    Paint initPaint() {
        Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setAntiAlias(true);
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeJoin(Paint.Join.ROUND);
        paint.setStrokeCap(Paint.Cap.ROUND);
        return paint;
    }

    @Override
    int getAlpha() {
        return 0xff;
    }

    @Override
    public void setColor(int color) {
        externalColor = color;
        super.setColor(0xffffffff);
    }

    @Override
    public int getColor() {
        return externalColor;
    }

    @Override
    public void onTouchDown(int x, int y) {
        startX = x;
        startY = y;
        path.moveTo(x, y);
        path.lineTo(x, y);

        prevX = x;
        prevY = y;

        drawPath(path);
    }

    @Override
    public void onTouchMove(int x, int y) {
        if (path.isEmpty()) {
            path.moveTo(prevX, prevY);
        }
        if (x == startX && y == startY) {
            path.lineTo(x + 0.1f, y);
        } else {
            path.lineTo(x, y);
        }

        prevX = x;
        prevY = y;

        drawPath(path);
    }

    @Override
    public void onTouchUp(int x, int y) {
        if (path.isEmpty()) {
            path.moveTo(prevX, prevY);
        }
        path.lineTo(x, y);

        drawPath(path);

        prevX = 0;
        prevY = 0;
    }

    @Override
    public void onDraw() {
        path.reset();
    }

}

File: app/src/main/java/com/tomclaw/drawa/draw/tools/Eraser.kt
package com.tomclaw.drawa.draw.tools

import android.graphics.Paint
import android.graphics.Path

class Eraser : Tool() {

    private var startX: Int = 0
    private var startY: Int = 0
    private var prevX: Int = 0
    private var prevY: Int = 0
    private var path = Path()

    override var color = ERASER_COLOR
    override val alpha = 0xff
    override val type = TYPE_ERASER

    override fun initPaint() = Paint().apply {
        style = Paint.Style.STROKE
        strokeJoin = Paint.Join.ROUND
        strokeCap = Paint.Cap.ROUND
        color = ERASER_COLOR
    }

    override fun onTouchDown(x: Int, y: Int) {
        resetRadius()

        startX = x
        startY = y

        path.moveTo(x.toFloat(), y.toFloat())
        path.lineTo(x.toFloat(), y.toFloat())

        prevX = x
        prevY = y

        drawPath(path)
    }

    override fun onTouchMove(x: Int, y: Int) {
        if (path.isEmpty) {
            path.moveTo(prevX.toFloat(), prevY.toFloat())
        }
        if (x == startX && y == startY) {
            path.lineTo(x + 0.1f, y.toFloat())
        } else {
            path.quadTo(
                    prevX.toFloat(),
                    prevY.toFloat(),
                    ((x + prevX) / 2).toFloat(),
                    ((y + prevY) / 2).toFloat()
            )
        }

        prevX = x
        prevY = y

        drawPath(path)
    }

    override fun onTouchUp(x: Int, y: Int) {
        if (path.isEmpty) {
            path.moveTo(prevX.toFloat(), prevY.toFloat())
        }
        path.quadTo(prevX.toFloat(), prevY.toFloat(), x.toFloat(), y.toFloat())

        path.reset()

        drawPath(path)

        prevX = 0
        prevY = 0
    }

    override fun onDraw() {}

}

const val ERASER_COLOR = -0x1
